### PR TITLE
FW-1870-Show group specific data in conversation section

### PR DESCRIPTION
**Pull request by Ojashwa**
### What do you want to achieve?
-

### PR Checklist

- [x] I have tested it locally and all functionalities are working fine.
- [ ] I have compared it with mocks and all design elements are the same.
- [ ] I have tested it in IE Browser.

### How was the code tested?

- **Start a conversation with installation script.**

image
----
- **Select the JavaScript context to _` Kommunicate widget iframe `_  as shown in the below image.**

image
----
- **After a conversation is get started try to send a metadata object in Kommunicate.updateConversationMetadata().**

image
----

- **Metadata can be shown here.**

image
-

- **Empty state for conversation info.**

image

- https://showgroupspecificdata.netlify.app

### What new thing you came across while writing this code? 
-

### In case you fixed a bug then please describe the root cause of it? 
-

NOTE: Make sure you're comparing your branch with the correct base branch

### CODE REVIEW

**Review by gargshantnu — APPROVED** [review]

File: webplugin/js/app/kommunicate.js [hunk 1]
```diff
@@ -116,6 +116,32 @@ $applozic.extend(true, Kommunicate, {
             }, SET_TIMEOUT_DURATION);
         }
     },
+    groupSpecificData: function (userSpecificMetadata) {
+        let type = typeof userSpecificMetadata;
+        let groupId = userSpecificMetadata.groupId;
+        let metadataToShow = JSON.stringify(userSpecificMetadata.metadata);
+        if (type == 'object' && groupId && metadataToShow) {
+            const groupDataResponse = Applozic.ALApiService.groupUpdate({
+                data: {
```
Comment (gargshantnu) [inline webplugin/js/app/kommunicate.js:125]: or ?
```js
KommunicateGlobal.window.
```

**Review by gargshantnu — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js [hunk 2]
```diff
@@ -118,27 +118,35 @@ $applozic.extend(true, Kommunicate, {
     },
     groupSpecificData: function (userSpecificMetadata) {
         let type = typeof userSpecificMetadata;
+        let metaType = typeof userSpecificMetadata.metadata;
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:121]: Where is it getting used?

File: webplugin/js/app/kommunicate.js [hunk 3]
```diff
@@ -118,27 +118,35 @@ $applozic.extend(true, Kommunicate, {
...
         let groupId = userSpecificMetadata.groupId;
         let metadataToShow = JSON.stringify(userSpecificMetadata.metadata);
         if (type == 'object' && groupId && metadataToShow) {
-            const groupDataResponse = Applozic.ALApiService.groupUpdate({
-                data: {
-                    groupId: groupId,
-                    metadata: {
-                        metadataToShow: metadataToShow,
+            if (typeof metadataToShow == 'object') {
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:125]: Metadata show if already there in above line

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js (hunk 2 above)
Comment (Ojashwa) [reply webplugin/js/app/kommunicate.js]: to check the type of metadata

File: webplugin/js/app/kommunicate.js [hunk 4]
```diff
@@ -116,6 +116,42 @@ $applozic.extend(true, Kommunicate, {
             }, SET_TIMEOUT_DURATION);
         }
     },
+    groupSpecificData: function (userSpecificMetadata) {
+        let [type, metaType, groupId,metadataToShow] = [
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:120]: use var in chat widget code, let might not work in IE
in dashboard code, it is fine to use let

File: webplugin/js/app/kommunicate.js [hunk 5]
```diff
@@ -116,6 +116,42 @@ $applozic.extend(true, Kommunicate, {
             }, SET_TIMEOUT_DURATION);
         }
     },
+    groupSpecificData: function (userSpecificMetadata) {
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:119]: rename function to updateConversationMetadata

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js [hunk 6]
```diff
@@ -116,6 +116,42 @@ $applozic.extend(true, Kommunicate, {
...
+        let [type, metaType, groupId,metadataToShow] = [
+            typeof userSpecificMetadata,
+            typeof userSpecificMetadata.metadata,
+            userSpecificMetadata.groupId,
+            JSON.stringify(userSpecificMetadata.metadata)
+        ];
+        if (type == 'object' && groupId && metadataToShow) {
+            if (metaType == 'object') {
+                const groupDataResponse = Applozic.ALApiService.groupUpdate({
+                    data: {
```
[the remark this answers is not in the document]
Comment (Prochnost55) [reply webplugin/js/app/kommunicate.js:129]: Call API for updating data only if the submitted data (i.e. userSpecificMetadata) meets the required set of rules.

**Review by Prochnost55 — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js [hunk 7]
```diff
@@ -116,6 +116,49 @@ $applozic.extend(true, Kommunicate, {
             }, SET_TIMEOUT_DURATION);
         }
     },
+    updateConversationMetadata: function (userSpecificMetadata) {
+        if (
+            userSpecificMetadata != null &&
+            userSpecificMetadata !== 'undefined'
+        ) {
+            var [groupMetadata, metadataToShow] = [
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:124]: what's the benefit of combining it one line using [. ]
this way it seems difficult to understand the code

Comment (Ojashwa) [reply webplugin/js/app/kommunicate.js]: updated

File: webplugin/js/app/kommunicate.js [hunk 8]
```diff
@@ -116,6 +116,48 @@ $applozic.extend(true, Kommunicate, {
             }, SET_TIMEOUT_DURATION);
         }
     },
+    updateConversationMetadata: function (userSpecificMetadata) {
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:119]: not able to understand why are you using 'user' ('userSpecificMetadata') in conversation related thing

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js (hunk 8 above)
Comment (Ojashwa) [reply webplugin/js/app/kommunicate.js:119]: updated to conversationMetadata

File: webplugin/js/app/kommunicate.js [hunk 9]
```diff
@@ -116,6 +116,48 @@ $applozic.extend(true, Kommunicate, {
...
+    updateConversationMetadata: function (conversationMetadata) {
+        if (
+            conversationMetadata != null &&
+            conversationMetadata !== 'undefined'
+        ) {
+            var groupMetadata = 
+                kommunicateCommons.isObject(conversationMetadata) &&
+                kommunicateCommons.isObject(conversationMetadata.metadata)
+                    ? conversationMetadata
+                    : false;
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:128]: i think setting false is not required, it should be null. verify once

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js [hunk 10]
```diff
@@ -116,6 +116,49 @@ $applozic.extend(true, Kommunicate, {
             }, SET_TIMEOUT_DURATION);
         }
     },
+    updateConversationMetadata: function (conversationMetadata) {
+        if (
+            conversationMetadata != null &&
+            conversationMetadata !== 'undefined'
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:122]: use typeof for undefined check

File: webplugin/js/app/kommunicate.js [hunk 11]
```diff
@@ -116,6 +116,49 @@ $applozic.extend(true, Kommunicate, {
...
+        ) {
+            conversationMetadata = JSON.parse(JSON.stringify(conversationMetadata));
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:124]: why is this required to first change JSON to stringify and then back to JSON?

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js [hunk 12]
```diff
@@ -116,6 +116,45 @@ $applozic.extend(true, Kommunicate, {
             }, SET_TIMEOUT_DURATION);
         }
     },
+    updateConversationMetadata: function (conversationMetadata) {
+        if (
+            conversationMetadata != null &&
+            typeof conversationMetadata !== 'undefined'
+        ) {
+            var _metadata = Object.assign({},conversationMetadata);
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:124]: What's the reason for using _ ?

File: webplugin/js/app/kommunicate.js [hunk 13]
```diff
@@ -116,6 +116,45 @@ $applozic.extend(true, Kommunicate, {
...
+            var metadataToShow = JSON.stringify(_metadata.metadata);
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:125]: if it is used only inside 'if' block then it should be declared inside if

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js (hunk 12 above)
Comment (Ojashwa) [reply webplugin/js/app/kommunicate.js:124]: we already have some declaration named `metadata`.

Comment (devashishmamgain) [reply webplugin/js/app/kommunicate.js:124]: where is it?

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js (hunk 12 above)
Comment (Ojashwa) [reply webplugin/js/app/kommunicate.js:124]: image
-
image

Comment (devashishmamgain) [reply webplugin/js/app/kommunicate.js]: with that logic, then above functions should also not work because it is already present in 1st function, then how come it is working on 2nd function as well?

File: webplugin/js/app/kommunicate.js [hunk 14]
```diff
@@ -116,6 +116,47 @@ $applozic.extend(true, Kommunicate, {
...
+                kommunicateCommons.isObject(metadataToSend.metadata) &&
+                metadataToSend.groupId &&
+                metadataToSend.metadata
+            ) {
+                var metadataToShow = JSON.stringify(metadataToSend.metadata);
+                const groupDataResponse = Applozic.ALApiService.groupUpdate({
+                    data: {
+                        groupId: metadataToSend.groupId,
+                        metadata: {
+                            metadataToShow: metadataToShow,
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:136]: shall we rename it to conversationMetadata?
then it will be consistent naming everywhere docs, code, and dashboard

File: webplugin/js/app/kommunicate.js [hunk 15]
```diff
@@ -116,6 +116,47 @@ $applozic.extend(true, Kommunicate, {
...
+                        },
+                    },
+                    success: function (response) {
+                        // console.log(response);
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:140]: remove console.log commented code

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js [hunk 16]
```diff
@@ -116,6 +116,45 @@ $applozic.extend(true, Kommunicate, {
             }, SET_TIMEOUT_DURATION);
         }
     },
+    updateConversationMetadata: function (conversationMetadata) {
+        if (
+            conversationMetadata != null &&
+            typeof conversationMetadata !== 'undefined'
+        ) {
+            var metadataToSend = Object.assign({}, conversationMetadata);
+            var metadataToShow = Object.assign({}, conversationMetadata.metadata);
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:125]: duplicate line

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js (hunk 16 above)
Comment (Ojashwa) [reply webplugin/js/app/kommunicate.js:125]: both are cloning the different objects which are coming into the argument.

**Review by devashishmamgain — APPROVED** [review]

File: webplugin/js/app/kommunicate.js [hunk 17]
```diff
@@ -116,6 +116,45 @@ $applozic.extend(true, Kommunicate, {
             }, SET_TIMEOUT_DURATION);
         }
     },
+    updateConversationMetadata: function (conversationMetadata) {
+        if (
+            conversationMetadata != null &&
+            typeof conversationMetadata !== 'undefined'
```
Comment (gargshantnu) [inline webplugin/js/app/kommunicate.js:122]: can we use:
```suggestion
            !conversationMetadata
```

File: webplugin/js/app/kommunicate.js [hunk 18]
```diff
@@ -116,6 +116,45 @@ $applozic.extend(true, Kommunicate, {
...
+        ) {
+            var metadataToSend = Object.assign({}, conversationMetadata);
+            var metadataToShow = Object.assign({}, conversationMetadata.metadata);
+            if (
+                kommunicateCommons.isObject(metadataToSend) &&
+                kommunicateCommons.isObject(metadataToShow) &&
+                metadataToSend.groupId &&
+                metadataToSend.metadata
+            ) {
```
Comment (gargshantnu) [inline webplugin/js/app/kommunicate.js:131]: can we directly unpack group id and metadata and then directly use that?
this code is a little confusing.

File: webplugin/js/app/kommunicate.js (hunk 17 above)
Comment (gargshantnu) [reply webplugin/js/app/kommunicate.js:122]: The purpose of this is to make code more easily readable.

**Review by Prochnost55 — APPROVED** [review]

File: .husky/pre-commit [hunk 19]
```diff
@@ -1,4 +0,0 @@
-#!/bin/sh
-. "$(dirname "$0")/_/husky.sh"
-
-npm run prettify-commit
```
[the remark this answers is not in the document]
Comment (Ojashwa) [reply .husky/pre-commit]: This was removed by mistake but the same was added in this PR.

**Review by gargshantnu — APPROVED** [review]

**Review by devashishmamgain — APPROVED** [review]

File: webplugin/js/app/kommunicate.js [hunk 20]
```diff
@@ -116,6 +116,47 @@ $applozic.extend(true, Kommunicate, {
...
     },
+    updateConversationMetadata: function (conversationMetadata) {
+        if (
+            conversationMetadata
+        ) {
+            var conversationInfo = Object.assign({}, conversationMetadata);
+            var metadataToSend = Object.assign(
+                {},
+                conversationMetadata.metadata
+            );
```
Comment (Prochnost55) [inline webplugin/js/app/kommunicate.js:127]: Object.assign is not supported in IE
Please check it in IE

**Review by Prochnost55 — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js [hunk 21]
```diff
@@ -116,6 +116,48 @@ $applozic.extend(true, Kommunicate, {
             }, SET_TIMEOUT_DURATION);
         }
     },
+    updateConversationMetadata: function (conversationMetadata) {
+        if (conversationMetadata) {
+            var conversationInfo = JSON.parse(
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:121]: what's the purpose of stringify and then parsing again?

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js (hunk 21 above)
Comment (Ojashwa) [reply webplugin/js/app/kommunicate.js:121]: So the purpose is to clone the object as this method is supported in ES5, and we are giving support to IE

Comment (devashishmamgain) [reply webplugin/js/app/kommunicate.js:121]: why do we need clone and not the original object itself?

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js (hunk 21 above)
Comment (Ojashwa) [reply webplugin/js/app/kommunicate.js:121]: the incoming parameter is not behaving like an object so I am cloning it to pass the object verification.

Comment (devashishmamgain) [reply webplugin/js/app/kommunicate.js:121]: why is the incoming parameter not behaving like an object?

Comment (Ojashwa) [reply webplugin/js/app/kommunicate.js:121]: is that okay if we pass JSONstring to the function `Kommunicate.updateConversationMetadata(JSON.stringify(object))` then parse it and use object further?

Comment (devashishmamgain) [reply webplugin/js/app/kommunicate.js:121]: i am not able to understand why a object passed from one place, can't be used at other place. please explain it with example

File: webplugin/js/app/kommunicate.js [hunk 22]
```diff
@@ -116,6 +116,44 @@ $applozic.extend(true, Kommunicate, {
...
         }
     },
+
+    updateConversationMetadata: function (conversationMetadata) {
+        if (conversationMetadata) {
+            var metadataToSend = conversationMetadata.metadata;
+            if (
+                conversationMetadata.groupId &&
+                metadataToSend &&
+                kommunicateCommons.isObject(conversationMetadata) &&
```
Comment (gargshantnu) [inline webplugin/js/app/kommunicate.js:126]: at 121 we are already checking if `conversationMetadata` exists, so maybe we can ignore this check
```js
kommunicateCommons.isObject(conversationMetadata) &&
```

Comment (Ojashwa) [reply webplugin/js/app/kommunicate.js:126]: This is an object verification check.

Comment (gargshantnu) [reply webplugin/js/app/kommunicate.js:126]: first, we are accessing a property assuming it is an object, here:
```js
var metadataToSend = conversationMetadata.metadata;
```
and later we are doing object validation.
```js
kommunicateCommons.isObject(conversationMetadata) &&
```
isn't little wrong and overhead?

File: webplugin/js/app/kommunicate.js [hunk 23]
```diff
@@ -116,6 +116,44 @@ $applozic.extend(true, Kommunicate, {
             }, SET_TIMEOUT_DURATION);
         }
     },
+
+    updateConversationMetadata: function (conversationMetadata) {
+        if (conversationMetadata) {
+            if (
+                conversationMetadata.groupId &&
+                conversationMetadata.metadata &&
+                kommunicateCommons.isObject(conversationMetadata) &&
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicate.js:125]: lets say if kommunicateCommons.isObject(conversationMetadata) returns false then will the code before it work fine? 
conversationMetadata.groupId &&
                conversationMetadata.metadata &&

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/kommunicate.js (hunk 23 above)
Comment (Ojashwa) [reply webplugin/js/app/kommunicate.js:123]: okay got it will update the order

**Review by gargshantnu — APPROVED** [review]

**Review by devashishmamgain — APPROVED** [review]

**Review by Prochnost55 — APPROVED** [review]
